packages/teleterm/src/preload.ts

```diff
@@ -25,27 +33,69 @@ const loggerService = createLoggerService({
 });
 
 Logger.init(loggerService);
+const logger = new Logger('preload');
 
 contextBridge.exposeInMainWorld('loggerService', loggerService);
 
-contextBridge.exposeInMainWorld('electron', getElectronGlobals());
+contextBridge.exposeInMainWorld(
+  'electron',
+  withErrorLogging(getElectronGlobals())
+);
 
 async function getElectronGlobals(): Promise<ElectronGlobals> {
   const [addresses, credentials] = await Promise.all([
     mainProcessClient.getResolvedChildProcessAddresses(),
     createGrpcCredentials(runtimeSettings),
   ]);
   const tshClient = createTshClient(addresses.tsh, credentials.tshd);
+  const startupClient = createStartupClient(addresses.tsh, credentials.tshd);
   const ptyServiceClient = createPtyService(
     addresses.shared,
     credentials.shared,
     runtimeSettings
   );
+  const { server: tshdEventsServer, resolvedAddress: tshdEventsServerAddress } =
+    await createTshdEventsServer(
+      runtimeSettings.tshdEvents.requestedNetworkAddress,
+      credentials.tshdEvents
+    );
+  const { service: tshdEventsService, subscribeToEvent: subscribeToTshdEvent } =
+    createTshdEventsService();
+
+  tshdEventsServer.addService(
+    apiService.TshdEventsServiceService,
+    // Whatever we use for generating protobufs generated wrong types.
+    // The types say that tshdEventsServer.addService expects an UntypedServiceImplementation as the
+    // second argument. ITshdEventsServiceService does implement UntypedServiceImplementation.
+    // However, what we actually need to pass as the second argument needs to have the shape of
+    // ITshdEventsServiceServer. That's why we ignore the error below.
+    // @ts-expect-error The generated protobuf types seem to be wrong.
+    tshdEventsService
+  );
+
+  // Here we send to tshd the address of the tshd events server that we just created. Then we send
+  // another request that is going to resolve once tshd prepares a client for that server and
+  // injects it into daemon.Service.
+  //
+  // All uses of tshClient must wait before waitForTshdEventsClient finishes. Otherwise we run into
+  // a risk of causing panics in tshd due to a missing tshd events client.
+  try {
+    await startupClient.resolveTshdEventsServerAddress(tshdEventsServerAddress);
+    await startupClient.waitForTshdEventsClient({ timeoutMs: 5_000 });
+  } catch (e) {
+    logger.error(e);
+    // Make sure the UI shows an understandable error and not just something like
+    // "DEADLINE_EXCEEDED: Deadline exceeded".
+    throw new Error(
+      `Ran into a problem while setting up the tshd events client: ${e.message}`
+    );
+  }
```

Comment:
> (…) it seems to me that we mix different level of abstractions. We have both high level functions like `createPtyService` and low level (…)
>
> What would you say for simplifying it to only two calls?

I think that makes sense. It'll obscure a bit how the actual server setup works, that tsh client shouldn't be used until `waitForTshdEventsClient` returns a response, and that it's going to timeout within 5 seconds but I think we can add a comment explaining that.

Reply:
@gzdunek The overall logic got simpler and the code in preload.ts got simpler too so I wonder how do you feel about its current state.

Now we just create the server, create the service, add service to the server and update the address in tshd. Do you think it still warrants moving it to a separate function?

I like how `getElectronGlobals` works as this orchestrator of whatever needs to be set up during preload using a bunch of small functions. I think moving the tshd events stuff to a separate function wouldn't help here much and at best obscure the logic somewhat.

Before we had these extra RPC calls, startup client, extra error handling but now I think it's just much simpler.

Reply:
I'd consider moving `tshdEventsServer.addService()` into `createTshdEventsServer()` (IMO adding a service can be considered as a part of server creation). It would reduce the number of returned values. 
But I don't have any strong opinion on it, we can stay with the current version, it's simple anyway.

Reply:
Done. Tbh it neatly shows that the only things the outside world needs from `createTshdEventsServer` is the resolved address of the server and `subscribeToTshdEvent`.